subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/configurations/model/ConfigurationReportModelFactory.java

```diff
@@ -0,0 +1,196 @@
+        final List<ReportAttribute> attributesWithCompatibilityRules = scanner.getAttributesWithCompatibilityRules();
+        final List<ReportAttribute> attributesWithDisambiguationRules = scanner.getAttributesWithDisambiguationRules();
+
+        final Map<String, ReportConfiguration> convertedConfigurations = new ConcurrentHashMap<>(project.getConfigurations().size());
+        project.getConfigurations().stream()
+            .map(ConfigurationInternal.class::cast)
+            .forEach(configuration -> getOrConvert(configuration, project, convertedConfigurations));
+
+        return new ConfigurationReportModel(project.getName(),
+            convertedConfigurations.values().stream().sorted(Comparator.comparing(ReportConfiguration::getName)).collect(Collectors.toList()),
+            attributesWithCompatibilityRules, attributesWithDisambiguationRules);
+    }
+
+    private ReportConfiguration getOrConvert(ConfigurationInternal configuration, Project project, Map<String, ReportConfiguration> convertedConfigurations) {
+        return convertedConfigurations.computeIfAbsent(configuration.getName(), name -> {
+            final List<ReportConfiguration> extendedConfigurations = new ArrayList<>(configuration.getExtendsFrom().size());
+            configuration.getExtendsFrom().stream()
+                .map(ConfigurationInternal.class::cast)
+                .forEach(c -> extendedConfigurations.add(getOrConvert(c, project, convertedConfigurations)));
+
+            return convertConfiguration(configuration, project, fileResolver, extendedConfigurations);
+        });
+    }
+
+    private ReportConfiguration convertConfiguration(ConfigurationInternal configuration, Project project, FileResolver fileResolver, List<ReportConfiguration> extendedConfigurations) {
+        // Important to lock the config prior to extracting the attributes, as some attributes, such as TargetJvmVersion, are actually added by this locking process
+        configuration.preventFromFurtherMutation();
+
+        final List<ReportAttribute> attributes = configuration.getAttributes().keySet().stream()
+            .map(a -> convertAttributeInContainer(a, configuration.getAttributes()))
+            .sorted(Comparator.comparing(ReportAttribute::getName))
+            .collect(Collectors.toList());
+
+        final List<ReportCapability> explicitCapabilities = configuration.getOutgoing().getCapabilities().stream()
+            .map(this::convertCapability)
+            .sorted(Comparator.comparing(ReportCapability::toGAV))
+            .collect(Collectors.toList());
+        final List<ReportCapability> capabilities;
+        if (explicitCapabilities.isEmpty()) {
+            capabilities = Collections.singletonList(convertDefaultCapability(project));
+        } else {
+            capabilities = explicitCapabilities;
+        }
+
+        final List<ReportArtifact> artifacts = Collections.unmodifiableList(configuration.getAllArtifacts().stream()
```

Comment:
Rather than collect to list and wrap it, we have Guava -- we can use `ImmutableList.toImmutableList()` as the collector.

Reply:
I've heard some conflicting advice on whether or not we want to expand our usage of Guava shortcuts for functionality already present in the JDK.  @big-guy - What do you think?

Reply:
Okay, I'll switch to the immutable collections then, thanks for clarifying.